native/cocos/renderer/gfx-base/GFXBuffer.h

```diff
@@ -43,10 +43,34 @@ class CC_DLL Buffer : public GFXObject, public RefCounted {
     void resize(uint32_t size);
     void destroy();
 
+    template <typename T>
+    struct StructuredWriter {
+        T *operator->() {
+            return ptr;
+        }
+        T *ptr;
+    };
+
+    template <typename T>
+    StructuredWriter<T> createWriter() {
+        return StructuredWriter<T>{reinterpret_cast<T*>(getStagingAddress())};
+    }
+
+    template <typename T>
+    void write(const T& value, uint32_t offset) const {
+        write(reinterpret_cast<const uint8_t*>(&value), offset, sizeof(T));
+    }
+
+    void write(const uint8_t *value, uint32_t offset, uint32_t size) const;
+
     virtual void update(const void *buffer, uint32_t size) = 0;
 
+    virtual void flush(const void *buffer) { update(buffer, _size); }
```

Comment:
Where is flush called? I didn't find the calling place in this pr.

Reply:
I forgot to remove the interface. No need to call flush. I will delete it.

Reply:
flush should be called in `Buffer::update`

Reply:
`Buffer::flush` and `Buffer::getStagingAddress` is protected now.